universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,199 @@
+}
+
+NamedValueRefManager::iterator NamedValueRefManager::begin() const {
+    CheckPendingRefs();
+    return m_value_refs.begin();
+}
+
+NamedValueRefManager::iterator NamedValueRefManager::end() const {
+    CheckPendingRefs();
+    return m_value_refs.end();
+}
+
+NamedValueRefManager& NamedValueRefManager::GetNamedValueRefManager() {
+    ErrorLogger() << "NamedValueRefManager::GetNamedValueRefManager starts in process " << ::getpid() << " thread: " << std::this_thread::get_id();
+    static NamedValueRefManager manager; // function local 
+    ErrorLogger() << "NamedValueRefManager::GetNamedValueRefManager at " << &manager << " in process " << ::getpid() << " thread: " << std::this_thread::get_id();
+    return manager;
+}
+
+
+unsigned int NamedValueRefManager::GetCheckSum() const {
+    CheckPendingRefs();
+    unsigned int retval{0};
+    /*    for (auto const& name_type_pair : m_value_refs)
+        CheckSums::CheckSumCombine(retval, name_type_pair);  // TODO FIXME
+    CheckSums::CheckSumCombine(retval, m_value_refs.size());
+
+
+    DebugLogger() << "NamedValueRefManager checksum: " << retval;*/
+    return retval;
+}
+
+
+template <typename T>
+std::string NamedValueRefManager::RegisterValueRef(std::string valueref_name, std::unique_ptr<T> vref) {
+    InfoLogger() << "Register valueref for " << valueref_name << ": " << vref->Description();
+    if (m_value_refs.count(valueref_name)>0) {
+        ErrorLogger() << "Skip registration for already registered valueref for " << valueref_name;
+        ErrorLogger() << "Number of registered ValueRefs: " << m_value_refs.size();
+        return valueref_name;
+    }
+    m_value_refs.emplace(valueref_name, std::move(std::unique_ptr<ValueRef::AnyValueRef>(std::move(vref))));
+    ErrorLogger() << "Number of registered ValueRefs: " << m_value_refs.size();
+    return valueref_name;
+}
+
+
+NamedValueRefManager& GetNamedValueRefManager()
+{ return NamedValueRefManager::GetNamedValueRefManager(); }
+
+ValueRef::AnyValueRef* GetAnyValueRef(const std::string& name)
+{
+    InfoLogger() << "NamedValueRefManager::GetAnyValueRef look for registered valueref for \"" << name << '"';
+    if ( GetNamedValueRefManager().GetAnyValueRef(name) )
+        return GetNamedValueRefManager().GetAnyValueRef(name);
+    InfoLogger() << "NamedValueRefManager::GetAnyValueRef could not find registered valueref for \"" << name << '"';
+    return nullptr;
+}
+
+template <typename T>
+ValueRef::ValueRef<T>* GetValueRef(const std::string& name)
+{ return GetNamedValueRefManager().GetValueRef<T>(name); }
+
+template <>
+ValueRef::ValueRef<int>* GetValueRef(const std::string& name)
+{ return GetNamedValueRefManager().GetValueRef<int>(name); }
+
+template <>
+ValueRef::ValueRef<double>* GetValueRef(const std::string& name)
+{ return GetNamedValueRefManager().GetValueRef<double>(name); }
+
+template <typename T>
+std::string RegisterValueRef(std::string name, std::unique_ptr<T>&& vref) {
```

Comment:
do you use the returned string from `RegisterValueRef` for anything?

Reply:
I wanted to use the given name parameter (when it was still a valueref reference) in the parser for creating the NamedRef in the name-in-the-middle case . I could not assign it to a lazy variable in a way that i could use it for constructing the enveloped NamedRef. 
So this can be void now.